client/state/memberships/coupon-list/test/actions.js

```diff
@@ -34,4 +57,415 @@ describe( 'actions', () => {
 			} );
 		} );
 	} );
+
+	describe( 'requestAddCoupon()', () => {
+		test( 'should dispatch add coupon failure and error notice on failure', async () => {
+			const siteId = 1;
+			const coupon = {
+				coupon_code: 'COUPON4',
+				discount_type: COUPON_DISCOUNT_TYPE_PERCENTAGE,
+				discount_value: 50,
+				discount_percentage: 50,
+				start_date: '2023-12-20',
+				end_date: undefined,
+				plan_ids_allow_list: [],
+				cannot_be_combined: true,
+				can_be_combined: false, //TODO: remove when backend switches to cannot_be_combined
+				first_time_purchase_only: false,
+				duration: COUPON_DURATION_FOREVER,
+				email_allow_list: [],
+			};
+			const errorReason = 'coupon already exists';
+			const errorNoticeText = 'Error adding coupon: ' + errorReason;
+			const noticeText = 'Added coupon';
+			const dispatchedActions = [];
+
+			const dispatch = ( obj ) => dispatchedActions.push( obj );
+			jest
+				.spyOn( wpcom.req, 'post' )
+				.mockImplementation( () => Promise.reject( { message: errorNoticeText } ) );
+			await requestAddCoupon( siteId, coupon, noticeText )( dispatch );
+			expect( wpcom.req.post.mock.calls ).toEqual( [
+				[
+					{
+						method: 'POST',
+						path: '/sites/1/memberships/coupons',
+						apiNamespace: 'wpcom/v2',
+					},
+					coupon,
+				],
+			] );
+			const dispatchedActionsWithUniqueIdsRemoved = dispatchedActions.map( ( action ) => {
+				if ( action.notice?.noticeId?.length > 0 ) {
+					action.notice.noticeId = 'UNIQUE-ID';
```

Comment:
Mutating the `action` is really not great and can lead to fragile tests.

Reply:
This was done to remove one randomly generated value from the object before comparing. I've refactored and am now just pulling the randomly generated value from the dispatched action and putting it into the expectation.